straxen/plugins/position_reconstruction.py

```diff
@@ -0,0 +1,171 @@
+"""Position reconstruction for Xenon-nT"""
+
+import os
+import tempfile
+import tarfile
+import numpy as np
+import strax
+import straxen
+
+export, __all__ = strax.exporter()
+
+
+@export
+@strax.takes_config(
+    strax.Option('min_reconstruction_area',
+                 help='Skip reconstruction if area (PE) is less than this',
+                 default=10),
+    strax.Option('n_top_pmts', default=straxen.n_top_pmts,
+                 help="Number of top PMTs")
+)
+class PeakPositionsBaseNT(strax.Plugin):
+    """
+    Base class for reconstructions.
+    This class should only be used when subclassed for the different
+    algorithms. Provides x_algorithm, y_algorithm for all peaks > than
+    min-reconstruction area based on the top array.
+    """
+    depends_on = ('peaks',)
+    algorithm = None
+    # check that this works properly, significant loading/saving penalty
+    # for three plugins!
+    parallel = 'process'
+    __version__ = '0.0.0'
+
+    def infer_dtype(self):
+        if self.algorithm is None:
+            raise NotImplementedError(f'Base class should not be used without '
+                                      f'algorithm as done in {__class__.__name__}')
+        dtype = [('x_' + self.algorithm, np.float32,
+                  f'Reconstructed {self.algorithm} S2 X position (cm), uncorrected'),
+                 ('y_' + self.algorithm, np.float32,
+                  f'Reconstructed {self.algorithm} S2 Y position (cm), uncorrected')]
+        dtype += strax.time_fields
+        return dtype
+
+    def setup(self):
+        if self.config[f'file_{self.algorithm}'] is None:
+            raise ValueError(f'No file provided to reconstruction with for {self.algorithm}')
+        import tensorflow as tf
+        self.model_file = str(self.config[f'file_{self.algorithm}'])
+        self.model = None
+        if os.path.exists(self.model_file):
+            print(f"Path is local. Loading {self.algorithm} TF model locally "
+                  f"from disk.")
+        else:
+            downloader = straxen.MongoDownloader()
+            if self.model_file in downloader.list_files():
+                self.model_file = downloader.download_single(self.model_file)
+            else:
+                raise RuntimeError(f'Model files {self.model_file} is not found')
+        with tempfile.TemporaryDirectory() as tmpdirname:

```

Comment:
One thing I noticed is that we cannot pickle the `self.model` leading to some nasty pickle errors in multiprocessing. We should discuss the best solution, now I've done:

```python
  def setup(self):
        ...
        resource_dir = os.path.join(
            './resource_cache',
            strax.utils.deterministic_hash(self.model_file))
        if not os.path.exists(resource_dir):
            os.makedirs(resource_dir)
            tar = tarfile.open(self.model_file, mode="r:gz")
            tar.extractall(path=resource_dir)
        self.model_path = resource_dir

    def compute(self, peaks):
        import tensorflow as tf
        model = tf.keras.models.load_model(self.model_path)
        # self.model -> model
```

Reply:
That's the reason why we chose to save the model this way. TensorFlow/Keras has many not pickleable stuff (especially when custom layers/operations are involved).

I think this solution might make processing slower as it takes a few seconds to load the model. So probably a few more seconds for each chunk...

Reply:
I see that but you simply cannot add non-pickable objects to the Plugin class:
https://github.com/AxFoundation/strax/blob/master/strax/plugin.py#L45

Reply:
TensorFlow does not really support pickling custom objects:
https://www.tensorflow.org/guide/keras/save_and_serialize#custom_objects

Maybe the ultimate solution is switch to another framework. It looks like PyTorch models are pickable. 

I wonder do pickle errors also appear when only the MLP model is used?

Reply:
So it's a python multiprocessing/strax thing, during multiprocessing python pickles these objects so all three PosRecs plugins would be affected. I guess this is why Jelle saved the 1T model this way in two seperate files, would that be an option?

There are some hacky solutions but I'm not sure if we should go down this road, especially given the link you just provided:
http://zachmoshe.com/2017/04/03/pickling-keras-models.html

PyTorch could be an option, but it's quite a heavy dependency which would slow down straxen-testing a bit (still better than slowing straxen processing down). I agree that what I wrote down as an example is not a solution we want. Perhaps @terliuk or @tunnell has an idea on how to save the model in such a way that it allows pickling and therefore can be set as a class-attribute.

Reply:
Can we test the PyTorch option elsewhere? Does that change the PR very significantly later?

In another thread, I know that @reallsx can improve the speed by ~30% by playing with batch size. I don't know if that helps more for CNN. But I'm not sure if it's significant thing to consider now.